### PR TITLE
[iOS] Implement the nativeID functionality in a more efficient way

### PR DESCRIPTION
**Pull request by zhongwuzw**
## Summary

Implement TODO, implement the nativeID functionality in a more efficient way instead of searching the whole view tree.

## Changelog

[iOS] [Fixed] - Implement the nativeID functionality in a more efficient way

## Test Plan

Circle CI passes, and nativeID functionality works as before.

```
  constructor(props) {
    super(props);
    this.state = {text: 'Placeholder Text'};
  }

  render() {
    const inputAccessoryViewID = "uniqueID";
    return (
      <View>
        <ScrollView keyboardDismissMode="interactive">
          <TextInput
            style={{
              padding: 10,
              paddingTop: 50,
            }}
            inputAccessoryViewID={inputAccessoryViewID}
            onChangeText={text => this.setState({text})}
            value={this.state.text}
          />
        </ScrollView>
        <InputAccessoryView nativeID={inputAccessoryViewID}>
          <Button
            onPress={() => this.setState({text: 'Placeholder Text'})}
            title="Reset Text"
          />
        </InputAccessoryView>
      </View>
    );
  }
```

### CODE REVIEW

File: React/Modules/RCTUIManager.m [hunk 1]
```diff
@@ -366,31 +378,27 @@ - (void)setLocalData:(NSObject *)localData forView:(UIView *)view
...
- */
 - (UIView *)viewForNativeID:(NSString *)nativeID withRootTag:(NSNumber *)rootTag
 {
   RCTAssertMainQueue();
-  UIView *view = [self viewForReactTag:rootTag];
-  return [self _lookupViewForNativeID:nativeID inView:view];
+  if (!nativeID || !rootTag) {
+    return nil;
+  }
+  return [_nativeIDRegistry objectForKey:[NSString stringWithFormat:@"%@%@", nativeID, rootTag]];
```
[the remark this answers is not in the document]
Comment (zhongwuzw) [reply React/Modules/RCTUIManager.m]: @shergin 👍 Done.

**Comment by PeteTheHeat** [comment]
Hey @zhongwuzw I'm curious why you used <nativeID-rootViewTag> as the key for this map? Product code shouldn't duplicate nativeID, correct? 

I'm asking because I found a race condition in an internal test. It was caused because nativeIDRegistry is read before the completion block in setNativeID executes. I want to change the key to simply <nativeID>

**Comment by zhongwuzw** [comment]
@PeteTheHeat Yes, `nativeID` cannot duplicate in a single `RootView`, In previously, https://github.com/facebook/react-native/blob/0.58-stable/React/Modules/RCTUIManager.m#L373, we also only support one root view vs one unique nativeID, because we iterate the subviews to find.

If we want to get the view using nativeID, we can do like `RCTBaseTextInputView` does.
https://github.com/facebook/react-native/blob/7a7eb11965a4126987fd0f4bfcd13dd872452e76/Libraries/Text/TextInput/RCTBaseTextInputView.m#L550

**Comment by MikePendo** [comment]
I have a question about this implementation:
```
- (UIView *)viewForNativeID:(NSString *)nativeID withRootTag:(NSNumber *)rootTag
{
  if (!nativeID || !rootTag) {
    return nil;
  }
  UIView *view;
  @synchronized(self) {
    view = [_nativeIDRegistry objectForKey:RCTNativeIDRegistryKey(nativeID, rootTag)];
  }
  return view;
}

```
1.Where I suppose to get the rootTag?
2. According to the docs `@param rootTag the react tag of root view hierarchy from which to find the view.`
But form the code I see that u search the view based on `[NSString stringWithFormat:@"%@-%@", rootTag, nativeID]` from the Map so it doesnt really search the view  starting from some rootTag (assuming that I mange to find the _initial_ rootView).
So I am a little bit confused for how to get the view based on the nativeId only which I specify in JS.
 Thanks .

**Pull request closed without merging**